src/formatter.ts

```diff
@@ -0,0 +1,78 @@
+import * as cp from "child_process";
+import * as os from "os";
+import * as path from "path";
+import * as vscode from "vscode";
+import { TextEncoder } from "util";
+
+export class Formatter implements vscode.DocumentFormattingEditProvider {
+    readonly binaryPath: string = '';
+
+    constructor(binaryPath: string) {
+        if (!binaryPath || binaryPath.length === 0) {
+            throw new Error('binaryPath is required to construct a formatter');
+        }
+        this.binaryPath = binaryPath;
+    }
+
+    public provideDocumentFormattingEdits(
+        document: vscode.TextDocument,
+        // @ts-ignore: The formatter doesn't have any options, but we need to implement the interface.
+        options: vscode.FormattingOptions,
+        token: vscode.CancellationToken
+    ): vscode.ProviderResult<vscode.TextEdit[]> {
+        if (vscode.window.visibleTextEditors.every((e) => e.document.fileName !== document.fileName)) {
+            return [];
+        }
+        return this.runFormatter(document, token).then(
+            (edits) => edits,
+            (err) => {
+                console.log(err);
+                return Promise.reject('Check the console in dev tools to find errors when formatting.');
+            }
+        );
+    }
+
+    private randomId(): string {
+        return (Math.floor(Math.random() * 100000)).toString()
+    }
+
+    private runFormatter(
+        document: vscode.TextDocument,
+        token: vscode.CancellationToken
+    ): Thenable<vscode.TextEdit[]> {
+        return new Promise<vscode.TextEdit[]>((resolve, reject) => {
+            const cwd = path.join(os.tmpdir(), this.randomId());
+            vscode.workspace.fs.createDirectory(vscode.Uri.file(cwd)).then(() => {
```

Comment:
I know it's extremely unlikely (and probably overkill), but should we check if the directory exists before we proceed with the random ID we selected? We could continue to generate a new ID until we find one that isn't already occupied (stopping after a certain number of attempts so we don't try forever). This is similar to what [os.CreateTemp](https://cs.opensource.google/go/go/+/refs/tags/go1.18:src/os/tempfile.go;drc=refs%2Ftags%2Fgo1.18;l=30) does.

It seems like a simple enough function to build out, so it might be worth it.

Reply:
Fair point - I'd _try_ that, but my js-foo is way too subpar to do it cleanly. Given that the chances of collision are tiny, it's likely not worth it (particularly if reading from stdin is coming anytime soon)

Reply:
Agreed, I'm fine with leaving this as-is.